repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -70,29 +75,41 @@ class Session(interpreter: Interpreter)
 
   def state: SessionState = _state
 
-  def history: IndexedSeq[Statement] = _history
+  def statements: mutable.Map[Int, Statement] = _statements
+
+  def execute(code: String): Int = {
+    val statementId = newStatementId.getAndIncrement()
+    _statements.synchronized {
```

Comment:
Realistically speaking, even with HA, no one's gonna connect thousands of livy-server to the same livy-repl. This sounds like a premature optimization.